### DISCUSSION TITLE
_1_mask: What's its role?

### DISCUSSION THREAD

**Discussion by sirdody** [harfbuzz/harfbuzz#3751] [General] [answered]
Hello, 

Each feature has a mask (hb_mask_t), so what is the role of _1_mask? What's use for? 

Thank you

**Comment by behdad** [comment, marked as answer]
The _1_mask is used for global features, such that we don't waste a bit for each of the features that are on for all the glyphs. The _1_mask is on for all glyphs.

**Comment by sirdody** [comment]
Each glyph has its own feature mask (hb_glyph_info_t), but in feature_map_t has both mask and _1_mask. So what's the different between mask and _1_mask in feature_map_t?

**Reply by behdad** [reply]
Ah, you're right. My bad. I had confused _1_mask with global_mask.

As the comment says:
```
    hb_mask_t mask; 
    hb_mask_t _1_mask; /* mask for value=1, for quick access */ 
```
so it's just an optimization. It's value is just:
```
    map->_1_mask = (1u << map->shift) & map->mask; 
```

It's used to set masks on glyphs for a particular feature. For example, hb-ot-shape.cc saves the _1_mask for various features, to use them for setting glyph masks later.
